sdk/keyvault/keyvault-secrets/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -0,0 +1,138 @@
+import { Constants } from "@azure/core-http";
+import { HttpOperationResponse } from "@azure/core-http";
+import { HttpHeaders } from "@azure/core-http";
+import { WebResource } from "@azure/core-http";
+
+export const TokenRefreshBufferMs = 2 * 60 * 1000; // 2 Minutes
+
+/**
+ * Creates a new ChallengeBasedAuthenticationPolicy factory.
+ *
+ * @param credential The TokenCredential implementation that can supply the challenge token.
+ */
+export function challengeBasedAuthenticationPolicy(credential: TokenCredential): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new ChallengeBasedAuthenticationPolicy(nextPolicy, options, credential);
+    }
+  };
+}
+
+export class AuthenticationChallenge {
+  constructor(public scopes: string[] | string) { }
+}
+
+/**
+ *
+ * Provides a RequestPolicy that can request a token from a TokenCredential
+ * implementation and then apply it to the Authorization header of a request
+ * as a Bearer token.
+ *
+ */
+export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+  private cachedToken: AccessToken | undefined = undefined;
+  private challenge: AuthenticationChallenge | undefined = undefined;
+
+  /**
+   * Creates a new ChallengeBasedAuthenticationPolicy object.
+   *
+   * @param nextPolicy The next RequestPolicy in the request pipeline.
+   * @param options Options for this RequestPolicy.
+   * @param credential The TokenCredential implementation that can supply the bearer token.
+   * @param scopes The scopes for which the bearer token applies.
+   */
+  constructor(
+    nextPolicy: RequestPolicy,
+    options: RequestPolicyOptions,
+    private credential: TokenCredential,
+    private refreshOn: number = Date.now(),
+  ) {
+    super(nextPolicy, options);
+  }
+
+  private parseWWWAuthenticate(www_authenticate: string): string {
+    let authenticateArray = www_authenticate.split(" ");
+    delete authenticateArray[0];
+    let commaSep = authenticateArray.join().split(",");
+    for (let item of commaSep) {
+      let kv = item.split("=");
+      if (kv[0].trim() == "resource") {
+        let resource = kv[1].trim().replace(/['"]+/g, '');
+        return resource;
+      }
+    }
+    return "";
+  }
+
+  /**
+   * Applies the Bearer token to the request through the Authorization header.
+   * @param webResource
+   */
+  public async sendRequest(
+    webResource: WebResource
+  ): Promise<HttpOperationResponse> {
+    if (!webResource.headers) webResource.headers = new HttpHeaders();
+
+    let originalBody = webResource.body;
+
+    if (this.challenge == undefined) {
+      webResource.body = "";
+    } else {
+      await this.authenticateRequest(webResource);
+    }
+
+    // Use a blank to start the challenge
+    let headers = webResource.headers;
+    webResource.headers = new HttpHeaders();
+    webResource.body = "";
```

Comment:
Does this elicit a challenge for every request?

Reply:
Ah, good point. I thought I had fixed it to check for the existing challenge before starting one. Will fix.

Reply:
Circling back (I had to page back in what I did) - this uses the cached challenge for each request. So it runs: 

```ts
await this.authenticateRequest(webResource);
```

If there is a cached challenge, this will set up the headers for us.  Next, we send the request. If the request comes back and isn't a redirection, we just return the result (because we've now used the challenge to do the next policy in the chain)

Reply:
(saw the bug I think you were pointing out. I think I fixed it)